torch/quantization/fx/quantize.py

```diff
@@ -1317,11 +1318,9 @@ def record_match(pattern, node, matched):
                                 dtypes = get_qconfig_dtypes(this_node_qconfig)
                                 # TODO(future PR): update the pattern to quantize
                                 # handler logic to take this into account.
-                                skip_this_match = (
-                                    (node.target in binary_op_supported_dtypes) and
-                                    (dtypes not in binary_op_supported_dtypes[node.target])
-                                )
-
+                                skip_this_match = (node.target in binary_op_supported_dtypes and dtypes not in binary_op_supported_dtypes[node.target]) and not (
```

Comment:
@jerryzh168 , thoughts about making is_reference passed to prepare so we don't have to do hacks like this?  If is_reference is needed for proper placement of observers, it makes sense for prepare to know about it.

Reply:
not sure what is this guarding against, I saw both binary_op_supported_dtypes and binary_reference_op_supported_dtypes here, does passing is_reference help here?

Reply:
the original `skip_this_match` was a hack added by me.  Essentially the current matching logic only matches on the op pattern, and does not know about qconfig or is_reference while matching.  But, the actual logic needed is to match on the combination of pattern, qconfig for that pattern, and the value of is_reference.

For example, if `is_reference` was known at prepare time, this check could be written as

```
skip_this_match_if_not_is_reference = (
  (not is_reference) and
  (node.target in binary_op_supported_dtypes) and
  (dtypes not in binary_op_supported_dtypes[node.target])
)

skip_this_match_if_is_reference = (
  is_reference and
  node.target in binary_reference_op_supported_dtypes and 
  dtypes in binary_reference_op_supported_dtypes[node.target]
)

skip_this_match = skip_this_match_if_not_is_reference or skip_this_match_if_is_reference
```

Reply:
@HDCharles , could we go with something like this for the meantime for the sake of readability, as we probably will need to do the "when shoudl is_reference flag be set" discussion async

```
# note: the value of is_reference is unknown at prepare, so we have to cover both cases
# the first case is when is_reference = False
skip_this_match_if_not_is_reference = ( 
  (node.target in binary_op_supported_dtypes) and 
  (dtypes not in binary_op_supported_dtypes[node.target]) 
) 
# the first case is when is_reference = True
skip_this_match_if_is_reference = ( 
  (node.target in binary_reference_op_supported_dtypes) and 
  (dtypes in binary_reference_op_supported_dtypes[node.target])
) 
skip_this_match = skip_this_match_if_not_is_reference or skip_this_match_if_is_reference
```

Reply:
do you mean skip if both are true? what you have is going to gate anything not supported on both

Reply:
sure, I was more commenting on the readability, spelling out what exactly we know and don't know in this function, and clearly explaining what the condition is doing, keeping lines short, etc.

Reply:
For ops like bmm, whether or not we place observers depends on is_reference. We should therefore have prepare take this as an argument too. Can we follow up on this in a later PR?